Add support for the invisible ReCaptcha [MAILPOET-4145]

rodrigoprimo opened the pull request:
This PR adds support for the invisible ReCaptcha in the subscribe forms. Before, we only supported the checkbox ReCaptcha. The API is almost identical, but there are some differences. Because of those differences, it was necessary to change the code that handles the form submission on public.jsx a bit. 

For the invisible version, we need to call recaptcha.execute(), and then the ReCaptcha API calls the pre-defined callback set inside renderCaptcha() with a token if the captcha was solved successfully. The callback then calls submitSubscribeForm() with the token. For the checkbox version, we get the token synchronously inside the code that handles the form submission when recaptcha.getResponse() is called. So we can call submitSubscribeForm() directly.

I would suggest reviewing this PR commit by commit. The first commit makes the necessary changes to the admin to add a new setting to use the invisible ReCaptcha. The second commit just makes some organizational changes to public.jsx that are required for the next commit due to the no-use-before-define ESLint rule. Finally, the third commit changes the frontend code to display the invisible ReCaptcha if this is the selected setting.

**Testing instructions**

- Test that **submitting a form on the front-end works** for the three different types of captcha that we support (Built-in, Google reCAPTCHA v2 Checkbox, Google reCAPTCHA v2 Invisible) and also **when the captcha is disabled**.
- You will need to enter `Google ReCaptcha Keys` in `Settings` > `Advanced tab`... and find your keys in the `P2` doc
- Ensure that the **new setting in the admin works** according to the specs in the Jira ticket.
- It is a bit tricky to test that the invisible ReCaptcha is working as most of the time, it works without interfering with the form submission flow. You add a breakpoint to SubscriberSubscribeController::validateCaptcha() and confirm that this method is getting a ReCaptcha token and that the request that it makes to the ReCaptcha server has a successful response.
- Links to the ReCaptcha API documentation in the Jira ticket.

[MAILPOET-4145]

[MAILPOET-4145]: https://mailpoet.atlassian.net/browse/MAILPOET-4145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

triple0t commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -85,12 +155,31 @@ jQuery(($) => {
+      const params = { sitekey, size };
+
+      if (size === 'invisible') {
+        params.callback = function invisibleReCaptchaCallback(token) {
+          const formData = form.mailpoetSerializeObject() || {};
+          formData.data.recaptcha = token;
```
From what I understand, this function `invisibleReCaptchaCallback` will be executed by the recaptcha API as a callback. 
Would these variables `form`, `formData` be available within the function scope?.

I can see this error in the console
<img width="745" alt="Screenshot 2022-04-21 at 1 10 23 PM" src="https://user-images.githubusercontent.com/30554163/164455279-3377bfe5-abe6-4c90-ac6b-1e46a75320cf.png">

When using the Invisible ReCaptcha option, submitting the form and solving the challenge.

rodrigoprimo replied:
> From what I understand, this function invisibleReCaptchaCallback will be executed by the recaptcha API as a callback.

That is correct.

>  Would these variables form, formData be available within the function scope?.

I believe you are referring only to `form` as `formData` is defined inside the callback function. I'm not a JS specialist so I could be missing something here, but I believe the answer is that `form` is available within the callback function because of the way that local scope works in JS. A variable defined inside a function is only accessible from that function and any nested functions.

> I can see this error in the console

I don't see this error in my environment. I made some changes related to another problem that you mentioned in the review comment. Could you please test again and check if you still get it?

triple0t commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -299,88 +388,36 @@ jQuery(($) => {
         }
 
+        const formData = form.mailpoetSerializeObject() || {};
+        const size = form.children('.mailpoet_recaptcha').attr('data-size');
+
```
Could you reconfirm you are getting the right value for the `data-size` attribute here?

When using the Invisible ReCaptcha option, I was getting `size` as undefined when it should be 'invisible'

rodrigoprimo replied:
See https://github.com/mailpoet/mailpoet/pull/4084#issuecomment-1110270323

triple0t reviewed (commented):
Awesome work @rodrigoprimo 

I tested and I can confirm the checkbox reCaptcha option works

But I was unable to get the Invisible reCaptcha option to work.

Here is the error from the UI:
<img width="714" alt="Screenshot 2022-04-21 at 1 22 44 PM" src="https://user-images.githubusercontent.com/30554163/164457314-855c797d-924a-44e7-b0a3-cc18ee2bb3e5.png">

Data gotten from the server:
<img width="427" alt="Screenshot 2022-04-21 at 1 22 53 PM" src="https://user-images.githubusercontent.com/30554163/164457364-1dba8301-75c4-41b3-a910-b788c5829233.png">

POST request Form data:
<img width="491" alt="Screenshot 2022-04-21 at 1 23 11 PM" src="https://user-images.githubusercontent.com/30554163/164457435-e7933704-693a-4d5a-a20e-57f3c6ec38d2.png">

POST request response:
<img width="580" alt="Screenshot 2022-04-21 at 1 23 25 PM" src="https://user-images.githubusercontent.com/30554163/164457515-f82ccb41-c20a-444c-8f9b-df66b8eb4261.png">

Kindly check my comments.

Thank you

rodrigoprimo commented on mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx:
```diff
@@ -27,12 +58,25 @@ export default function Captcha() {
+              {(type === 'recaptcha' || type === 'recaptcha-invisible') && (
+                <span>{t('reCaptchaDescription')} </span>
+              )}
+              <a
+                className="mailpoet-link"
+                href="https://www.google.com/recaptcha/admin"
+                rel="noopener noreferrer"
+                target="_blank"
+              >
+                {t('signupForCaptchaKey')}
+              </a>
```
I'm inclined to agree with you on this, but this is not something that I changed in this PR. Before the changes proposed here, we were already showing the "Sign up for an API key pair here" link when "Disable" is selected. Maybe we should address this in a separate ticket or discuss it with Jan M. before proceeding with this change here?

pavel-mailpoet replied:
Yes please, only display the Sign up link when the reCAPTCHA is selected. The specification also asks to add:
"Please use the same reCAPTCHA type as in Google reCAPTCHA Settings." and I don't see that on the screenshot.

triple0t replied:
@pavel-mailpoet the "Please use the same reCAPTCHA type as in Google reCAPTCHA Settings." text shows up when any of the Google reCaptcha options is selected.
<img width="1577" alt="Screenshot 2022-04-28 at 8 26 45 AM" src="https://user-images.githubusercontent.com/30554163/165700533-0b2f598b-21b2-497d-a28b-9d87e0191488.png">

triple0t replied:
Hi @rodrigoprimo I'm Okay with the state of this PR. I'm going to re-assign it back to you. You can choose to move it ahead to QA or implement Pavels' suggestion.

Thank you 🙏

rodrigoprimo replied:
Thanks for your input, @pavel-mailpoet.

@triple0t, I added a new commit changing the logic that controls when the sign up link is visible. Now it should only appear when the user selects either ReCaptcha checkbox or ReCaptcha invisible. I'm assigning this PR back to you for a final review.

triple0t replied:
Approved. Thank you

rodrigoprimo commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -85,12 +155,31 @@ jQuery(($) => {
+
+    // Users should not be able to change the size if it is equal to 'invisible' as this would
+    // change the type of the ReCaptcha.
+    if (size !== 'invisible') {
+      size = Hooks.applyFilters('mailpoet_re_captcha_size', 'compact');
+    }
```
I don't know why we are using compact as the default value for the checkbox ReCaptcha. This is a decision that was made before this PR, and probably before I joined MailPoet. Do you reckon this is something that we should change?

triple0t replied:
The major difference I noticed was the size of the captcha.

**Compact**
<img width="734" alt="Screenshot 2022-04-27 at 4 15 08 PM" src="https://user-images.githubusercontent.com/30554163/165556353-411c4a30-ac93-4a09-9eef-5b9bd641c127.png">

**Normal**
<img width="723" alt="Screenshot 2022-04-27 at 4 20 37 PM" src="https://user-images.githubusercontent.com/30554163/165556457-637bc844-7fb2-4f3d-8bcf-3053be9ed722.png">

**No Captcha**
<img width="725" alt="Screenshot 2022-04-27 at 4 21 08 PM" src="https://user-images.githubusercontent.com/30554163/165556499-bd4fa854-95c4-428d-a6c1-de3545269785.png">


> This is a decision that was made before this PR, and probably before I joined MailPoet. Do you reckon this is something that we should change?

Without finding out the reason why we are using compact over normal, I believe we can leave it as it is.

rodrigoprimo commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -77,6 +77,76 @@ jQuery(($) => {
+        form.trigger('reset');
+        // reset validation
+        parsley.reset();
+        // reset captcha
+        if (window.grecaptcha && formData.recaptcha) {
+          window.grecaptcha.reset(formData.recaptcha);
+        }
```
I also noticed that we are calling window.grecaptcha.reset() twice and I'm not sure why. I opted to leave the code as is since this duplication was already present and, when creating this PR, I did not have time to investigate the reasons.

rodrigoprimo commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -299,88 +388,36 @@ jQuery(($) => {
+          } else {
+            formData.data.recaptcha = window.grecaptcha.getResponse(
+              formData.recaptcha,
+            );
```
I was confused as well by those two variables when I first read the original code to create this PR, but I failed to act on it. Thanks for nudging me to take action and for the great suggestions for the new names for `formData.recaptcha` and `formData.data.recaptcha`. I just pushed a new commit implementing this change (208aa6fbefc4f91ccd6400b3692a2203ac6d79d2).

triple0t replied:
Thank you for performing the update 🙌🏿

rodrigoprimo commented on mailpoet/assets/js/src/public.jsx:
```diff
@@ -85,12 +155,31 @@ jQuery(($) => {
+      if (size === 'invisible') {
+        params.callback = function invisibleReCaptchaCallback(token) {
+          const formData = form.mailpoetSerializeObject() || {};
+          formData.data.recaptcha = token;
```
See https://github.com/mailpoet/mailpoet/pull/4084#discussion_r859160836

rodrigoprimo commented:
Thanks for the detailed review, @triple0t!

I believe I answered all your comments and this PR is ready for a second review.

> But I was unable to get the Invisible reCaptcha option to work.

I completely missed that different form templates had different HTML structures. I was testing with a form template different than the one you used to test and that explains why it was working for me, but not for you. I pushed a new commit (35b22a706c21fae58d678a1d60fcfc1d117ba761) to make the parts of the code that rely on the HTML structure more lenient and now the invisible ReCaptcha should work for the form template that you are using.

triple0t reviewed (dismissed):
Tested and working well.

Well done @rodrigoprimo and Thanks for updating the method names. 👏🏾

triple0t reviewed (dismissed):
LGTM

rohankacharegs commented:
GlobalStep Verified

veljkho reviewed (changes requested):
Hi @rodrigoprimo 
I think we have issue with the new Invisible reCAPTCHa or at least not the good behavior...
It is the situation when the user receives some warning from reCAPTCHA v2 Invisible but unable to see it on the form itself... for example the warning "You need to wait X more minutes in order to subscribe again".. which is happening if you want to subscribe twice time in a row (even using different form) for a short amount of time (like 2 minutes).

Please check this GIF: https://d.pr/i/tyJteu
So, when clicking the Subscribe button nothing happens.. the error is missing here "you need to wait X more minutes..". On this page I have had the popup form where I was subscribed recently, and wanted to subscribe on this widget placed form also.. so probably the warning left on the popup form.

Is this fixable? Please let me know if you need more information regarding reproduction or  about the issue itself, would be glad to help.

rodrigoprimo commented:
Good catch, @veljkho. I was not able to reproduce exactly the same setup that you had in the screencast that you shared, but I was able to reproduce a very similar bug.

If I'm not missing something, the original code was not resetting the ReCaptcha when there was an error in the form submission. Due to differences between the checkbox and invisible ReCaptchas, this meant that for the former there was an error but it didn't block the form submission. For the latter, the error did block the form submission which is the problem that you found.

I added a new commit to make sure we reset the ReCaptcha when there is an error with the form submission and in my testing, it fixed the problem that you found (or at least a very similar one, since I couldn't reproduce exactly the same setup).

Could you please test again?

(I'm rerunning CircleCI as a flaky acceptance test that is not related to this PR failed)

veljkho reviewed (changes requested):
Hi Rodrigo!
I can still reproduce the issue. This is happening if you have more than 1 form on the same page... Please add 2 forms (can be 1 widget and 1 popup) on the same page and try subscribing to Popup first.. then reload the page and try subscribing to widget form..  you'll see that there's no warning message present, and that is because of the warning left on the popup instead showing on the widget form. Some conflicting is present here, the warnings left on previously submitted form. Could be this Google's issue?

rodrigoprimo commented:
@veljkho, I missed from your previous comment that you were testing with two different form types. So the problem that I fixed is likely not the same as the one that you found. Apologies for that. I will ping you to get access to your online environment so that I can see exactly what you are seeing. Thanks.

rodrigoprimo commented:
I chatted with @veljkho, and I understood the problem that he was referring to that I originally missed. What threw me off was that the screencast doesn't show that the problem happens when there are multiple forms on the page. Here is a screencast showing that when there are at least two forms on the page, and the invisible ReCaptcha is being used, the second form doesn't work as the code is wrongly submitting the first form:

https://user-images.githubusercontent.com/77215/167432283-c92a1ead-3482-468c-a862-5bc6602b5e2c.mp4

This was happening because of a bug in the code that I created that was always calling `grecaptcha.execute()` for the first form as I forgot to pass the ID of the ReCaptcha when calling this method. I missed this problem when testing because, while I did test with multiple forms on the same page, I tested only the first form, and this one works as the API uses its ID when no ID is passed to `grecaptcha.execute()`.

Here is a screencast of the second form working after the changes that I introduced in 1288929:

https://user-images.githubusercontent.com/77215/167432902-6d4c6deb-7aa9-4e72-b939-91af8bd5cd89.mp4

@veljkho, could you please check again? We might want @triple0t to check this PR one more time just to be on the safe side, as I introduced a few code changes after his review.

rohankacharegs commented:
GlobalStep Verified

veljkho commented:
It is working now @rodrigoprimo thanks! :)
Please fix conflicts so I can merge it.

rodrigoprimo commented:
@veljkho, I fixed the conflicts, but I had to use rebase instead of merge, and that means that I need another approval review before we can merge this PR. Could you please check it again?
